File: C++/Algorithms/Backtracking/N_Queen_Problem.cpp
#include <iostream>
using namespace std;

bool isSafe(int **arr, int x, int y, int n)
{
    //check for any queen in col
    for (int i = 0; i < x; i++)
    {
        if (arr[i][y] == 1)
            return false;
    }
    //left diagonal
    int row = x, col = y;
    while (row >= 0 && col >= 0)
    {
        if (arr[row][col] == 1)
            return false;
        row--;
        col--;
    }
    //right diagonal
    row = x;
    col = y;
    while (row >= 0 && col < n)
    {
        if (arr[row][col] == 1)
            return false;
        row--;
        col++;
    }
    return true;
}
//no need for col varialbe; we wont be checking for col, we will move to next row after placing the quen
bool nQueen(int **arr, int x, int n)
{
    if (x >= n)
    { //base case
        return true;
    }
    for (int col = 0; col < n; col++)
    {
        if (isSafe(arr, x, col, n))
        {
            arr[x][col] = 1;
            if (nQueen(arr, x + 1, n))
                return true;
            arr[x][col] = 0; //backtracking
        }
    }
    return false;
}
int main()
{
    int n;
    cin >> n;
    int **arr = new int *[n];
    for (int i = 0; i < n; i++)
    {
        arr[i] = new int[n];
        for (int j = 0; j < n; j++)
        {
            arr[i][j] = 0;
        }
    }
    if (nQueen(arr, 0, n))
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                cout << arr[i][j] << " ";
            }
            cout << "\n";
        }
    }
    return 0;
}

File: Java/Algorithms/Backtracking/NQueenProblem.java
import java.util.Scanner;

public class NQueenProblem {

	public static void main(String[] args) {
        
		Scanner sc = new Scanner(System.in);
		int row = 0;
		System.out.println("Enter number rows: ");
		int N = sc.nextInt();
		int[][] board = new int[N][N];
		solveNQueen(board, row, N);
		System.out.println("Total possible ouptut = " + count);
		sc.close();

	}
	static int count = 0;
	public static boolean solveNQueen(int[][] board, int row, int N) {
		if(row == N) {				//Base case
			for(int i = 0; i < N; i++) {
				for(int j = 0; j < N; j++) {
					if(board[i][j] == 1)
						System.out.print("Q ");
					else System.out.print("_ ");
				}
				System.out.println();
			}
			count++;
			System.out.println();
			System.out.println();
			
			return false;      //result is false so that we can get all possible outputs  
		}
		
		for(int col = 0; col < N; col++) {    //traverse through each column
			if(isSafe(board, row, col, N)) {    //checks for cell
				board[row][col] = 1;
				
				if(solveNQueen(board, row+1, N))   //checks for next row
					return true;
				board[row][col] = 0;      //backtracking
			}	
		}
		
		return false;
	}
	
	public static boolean isSafe(int[][] board, int row, int col, int N) {
		for(int i = 0; i < row; i++) 
			if(board[i][col] == 1)    //checks for upper rows
				return false;
		
		for(int i = row, j = col; i >= 0 && j < N; i--, j++)
			if(board[i][j] == 1)     //checks for upper right diagonal
				return false;
		
		for(int i = row, j = col; i >= 0 && j >= 0; i--, j--)
			if(board[i][j] == 1)     //checks for upper left diagonal
				return false;
		
		return true;       	//if all loop gets terminated without any result then return true
			
	}

}
